### PR TITLE
support cmake LINK_LIBRARY:WHOLE_ARCHIVE

### PR DESCRIPTION
**Pull request by eagleoflqj**
Related: https://discourse.cmake.org/t/error-when-crosscompiling-with-whole-archive-target-link/9394

whole-archive support needs to be explicitly declared in toolchain file, see cmake official [GNU](https://github.com/Kitware/CMake/blob/6be01c932e077306ebea15cef4b8befb29c5130e/Modules/Platform/Linker/GNU.cmake#L35-L44).

Test:

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.24)

project(whole)

add_library(whole_lib STATIC lib.cpp)
add_executable(whole main.cpp)
target_link_libraries(whole $<LINK_LIBRARY:WHOLE_ARCHIVE,whole_lib>)
```

lib.cpp
```cpp
#include <iostream>

static void init(void) __attribute__((constructor));
static void init(void) {
    std::cout << "init" << std::endl;
}
```

main.cpp
```cpp
#include <iostream>

int main() {
    std::cout << "main" << std::endl;
}
```

With latest emsdk, `emcmake cmake -B build` complains
```
configure: cmake -B build -DCMAKE_TOOLCHAIN_FILE=/path/to/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR=/Users/liumeo/github/emsdk/node/20.18.0_64bit/bin/node
-- Configuring done (2.2s)
CMake Error at CMakeLists.txt:6 (add_executable):
  Feature 'WHOLE_ARCHIVE', specified through generator-expression
  '$<LINK_LIBRARY>' to link target 'whole', is not supported for the 'CXX'
  link language.
```

After overriding emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake, it passes, and `cmake --build build` passes, `node build/whole.js` outputs
```
init
main
```
as expected.

### CODE REVIEW

**Review by sbc100 — COMMENTED** [review]
This LGTM, but can you add a new test to `test_cmake` in `test_other.py` that uses this feature?

File: test/cmake/whole_archive/main.cpp [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#include <iostream>
+
+int main() {
+    std::cout << "main" << std::endl;
```
Comment (sbc100) [inline test/cmake/whole_archive/main.cpp:4]: Can you use 2 space indentation?

File: test/cmake/whole_archive/lib.cpp [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#include <iostream>
+
+static void init(void) __attribute__((constructor));
+static void init(void) {
```
Comment (sbc100) [inline test/cmake/whole_archive/lib.cpp:4]: I think you can combine these into a single line:

```
__attribute__((constructor)) void init(void) {
   ...
}
```

File: test/cmake/whole_archive/lib.cpp [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+    std::cout << "init" << std::endl;
```
Comment (sbc100) [inline test/cmake/whole_archive/lib.cpp:5]: Can you use plain C and printf here to keep this as simple as possible?

**Review by sbc100 — APPROVED** [review]
lgtm % nits

**Review by sbc100 — APPROVED** [review]

File: test/cmake/whole_archive/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.24)
```
Comment (sbc100) [inline test/cmake/whole_archive/CMakeLists.txt:1]: It looks like we still run the CI tests with 3.16.3.

What would happen if you lowered this number to 3.16?  Would the WHOLE_ARCHIVE thing below simply not work?

I guess we need some way to skip this new test when an older version of cmake is found.  Normally we do this via environment variables such as `EMTEST_SKIP_XXX`.  In this case maybe something like `EMTEST_SKIP_NEW_CMAKE`?   And then set that in `.circleci/config.yaml`?

Comment (eagleoflqj) [reply test/cmake/whole_archive/CMakeLists.txt:1]: > Would the WHOLE_ARCHIVE thing below simply not work?

Yeah I'm afraid it's a [3.24+ only](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#link-features) feature.

> something like EMTEST_SKIP_NEW_CMAKE

If we use an env var to toggle, we need a way to distinguish `whole_archive` from other parameterized tests within the group of `test_cmake`. We can do it by adding a new parameter "cmake_version". But then we may not need the env var at all, as the actual cmake version can be determined by parsing `cmake --version`. Further more, the `cmake_version` may not be necessary as well, as version requirement can be read from `CMakeLists.txt` as well.
I'd like a way that adds least maintenance burden to you.

Comment (sbc100) [reply test/cmake/whole_archive/CMakeLists.txt:1]: I think you can do:

```
if test_dir == 'whole_archive' and 'EMTEST_SKIP_NEW_CMAKE' in os.environ:
   self.skipTest('EMTEST_SKIP_NEW_CMAKE set')
```

Comment (sbc100) [reply test/cmake/whole_archive/CMakeLists.txt:1]: The reason we don't do things like `--version` checking is that we don't want tests that implicitly skip based on the environment.  We want to folks (and bots) to have the explicitly ops out of certain tests.

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
Yes we need to add `EMTEST_SKIP_NEW_CMAKE` there too.

**Comment by dschuff** [comment]
I guess the other option would be to just update the install of CMake we have on those builders.

**Comment by sbc100** [comment]
> I guess the other option would be to just update the install of CMake we have on those builders.

Maybe .. although then we loose coverage of older cmake versions

**Pull request merged**